### ISSUE TITLE
'within' WebDriverIO selectors not being re-evaluated after DOM mutation

### ISSUE THREAD

**Issue by cameronbraid** [Codeception/CodeceptJS#637]
#### What are you trying to achieve?

I wish to wait for a DOM update to complete, waiting until the element is found in the DOM

pseudo code 
```
within('.myWidget', ()=>{

  I.executeScript(function() {
      $('.someDiv').click(); // this causes .myWidget to be replaced with a new element
  });

  I.waitForElement('.newElementInTheWidget")

})
```

I see that the within locator isn't being re-evaluated when the inner waitForElement is being evaluated.  

 i.e. i'd like the waitForElement('.newElementInTheWidget") to evaluate the selector ".myWidget .newElementInTheWidget" as per the within declaration

**Comment by APshenkin** [comment]
@cameronbraid Hi!
What version of codeceptJS do you use? We fixed this issue in 1.0.0 release

**Comment by cameronbraid** [comment]
codeceptjs-webdriverio 1.1.0

I'll work on a minimal standalone example and that will either show that I have done something wrong, or show up a bug :)

Cheers

Cameron

**Comment by APshenkin** [comment]
@cameronbraid I think I got the cause of the problem. I did not look very carefully at your code. When you change `.myWidget` with a new element you can't use `within('.myWidget', ()=>{` because it's a new element. So I think that the best way is to move your script executing before using `within`.

I don't think that it is good idea to re-evaluate `within` element after each step. It's will require some time and also it's not clean for test (You do several operations with one element first, but after some "magic" your next steps belong to new element)

**Comment by cameronbraid** [comment]
Yeah, that's what I mean - re-evaulating the within selector

Conceptually, when writing tests, I don't care if the elements are swapped out underneath.. all that matters to me is that when I click based on a selector (.searchForm .searchButton) that it works..  For example if the web dev is using a reactive style template language, like markojs, I don't necessarily know when elements get created/destroyed.. and in fact I don't care as long as there is a button .searchButton within a .searchForm that I can click on.  Even if the view layer created all elements afresh each go around of the event loop, my tests should still work as they are 'selector' based.

So, I guess I am arguing that conceptually I am writing tests against selectors not elements.  The fact that an element becomes stale should be an indicator to the testing system that it needs to retry evaluating the selector to see if a new element has been created.  

So in conclusion - could the 'within' method detect stale elements and re-evaluating on demand (recursively up the stack) when needed.  That would solve the performance issue of not re-evaluating for each interaction, and it would solve my issue :)

Food for thought

Cameron

**Comment by DavertMik** [comment]
Hey, @cameronbraid we discussed this issue with @APshenkin and we can't come to the good implementation for it. Techincally it means that after each step we will need to require to update the element even if it's not stale. So querying element after each step would reduce the execution speed and will add complexity to the code. 

So advice would be - use stable elements for within blocks. For most of cases that's where you need them: to test a widget on a page, to test something inside popup, i.e. test inside a fixed container.

Sorry, we can't have this implemented. If you would come with a patch for this feature we probably might accept it but we won't implement it by ourselfs because of reasons explained

**Comment by vNh123** [comment]
#667
